python/pyarrow/parquet.py

```diff
@@ -646,13 +656,19 @@ def _should_silently_exclude(self, file_name):
                 file_name in EXCLUDED_PARQUET_PATHS)
 
     def _visit_directories(self, level, directories, part_keys):
+        futures_list = []
         for path in directories:
             head, tail = _path_split(path, self.pathsep)
             name, key = _parse_hive_partition(tail)
 
             index = self.partitions.get_index(level, name, key)
             dir_part_keys = part_keys + [(name, index)]
-            self._visit_level(level + 1, path, dir_part_keys)
+            future = self._thread_pool.submit(self._visit_level,
```

Comment:
Is `_visit_level` even thread-safe?

Reply:
I believe it should be. It mutates `pieces` which is a python list by using `extend` which should be thread safe thanks to the GIL.